### PR TITLE
Implementation of Index Calculus for discrete_log

### PR DESCRIPTION
**Pull request by teschlg**
#### Brief description of what is fixed or changed
This patch adds "Index Calculus" to discrete_log, which is faster for larger numbers.
E.g. on my machine it takes about 12 seconds to handle 16 digit numbers in comparison
to 51 seconds without the patch.

#### Other comments
One could further speed up things by searching for the relations in parallel.

#### Release Notes

* ntheory
  * Added index calculus as a new method for discrete_log

### CODE REVIEW

**Comment by haru-44** [comment]
I will check back over the weekend for more details, but I would appreciate it if you could fix the same as #26487 regarding coding style.

**Comment by haru-44** [comment]
Let's start with the basics.

## `_discrete_log_is_smooth`

### New line

```python
    """Try to factor n with respect to a a given factorbase. Upon success a list of exponents with repect to the factorbase is returned. Otherwise None."""
```

Please insert appropriate line breaks.

### Use of `enumerate`

```python
    for i in range(len(factorbase)):
        p=factorbase[i]
```

This can be rewritten as follows

```python
    for i, p in enumerate(factorbase):
```

### How to write trial

```python
        while ( n != 1 and n % p == 0 ): # divide by p as many times as possible
            factors[i] +=1
            n = n // p
```

`sympy.ntheory.factor_. _trial` could be used as a reference.

### New line

```python
    if n != 1: return None # the number factors if at the end nothing is left
```

This should be a new line.

### summary

After applying the above modifications, we would have the following

```python
def _discrete_log_is_smooth(n: int, factorbase: list):
    """Try to factor n with respect to a a given factorbase.
    Upon success a list of exponents with repect to the factorbase is returned.
    Otherwise None.
    """
    factors = [0]*len(factorbase)
    for i, p in enumerate(factorbase):
        if n % p == 0:
            n, m = remove(n // p, p)
            factors[i] = m + 1
    if n != 1:
        return None # the number factors if at the end nothing is left
    return factors
```

### Idea

Each time, we ensure `factors = [0]*len(factorbase)`, but if the probability of being smooth seems low, it may be faster to make a list after discovery.

**Comment by haru-44** [comment]
## `_discrete_log_index_calculus`

### How to find `B`
I did not understand the process of obtaining `B` very well. Could you please explain it in more detail?

### Comparison of `None`
You should use `is` instead of `==` to determine if an object is `None`.
`relations[i] ! = None` should be written as `relations[i] is not None` or `relations[i]`.

### isprime

```python
assert isprime(order), "The order of the base must be prime."
```

Since `order` is a private function that is assumed to be prime, there is no need to check it (although it is important to specify this in the documentation). It is not necessary to accept `prime_order` in the first place.

```python
def _discrete_log_index_calculus(n, a, b, order):
```

### Deep indentation
We do not want to make the indentation as deep as possible. In this case, two places can be fixed.

```python
        relation = _discrete_log_is_smooth(pow(b,x,n), factorbase)
        if relation:
```

```python
        relation = _discrete_log_is_smooth(pow(b,x,n), factorbase)
        if relation is None:
            continue
```

```python
            if index == lf or relations[index] != None: # the relation contains no new information
                pass
```

```python
            if index == lf or relations[index] != None: # the relation contains no new information
                continue
```

### Unused variables
I think `k` is essentially underutilized.


### Other
There should be two blank lines between functions.

**Comment by teschlg** [comment]
I don't get the line
```
n, m = remove(n // p, p)
```
I tried
```
n, rem = divmod(n, p)
```
from factor_trial/multiplicity, but that seems to make the code slower.

I have added a reference for how to obtain B.

**Comment by haru-44** [comment]
> I don't get the line
> 
> ```
> n, m = remove(n // p, p)
> ```
> 
> I tried
> 
> ```
> n, rem = divmod(n, p)
> ```
> 
> from factor_trial/multiplicity, but that seems to make the code slower.

I see. Then let's just go with it.
But is there a high probability that `n` will be smooth? If it is high, it would be better to use `n ! = 1` rather than checking `if n == 1: break` every time. Maybe the probability is low, so we omit `n ! = 1` can be omitted from the check.

> I have added a reference for how to obtain B.

Thanks, I wasn't sure where the 0.7 was coming from, but 0.5 is correct (Newton's method is probably wrong too).
While several approximate values are used to find the optimal `B`, there is not much advantage to be exact only for `u`. Thus, I see no problem in finding `B` directly.

> The `prime_order` test is just how it is done in the other functions. I just tried to be consistent here.

Consistent with what?
`_discrete_log_shanks_steps` and `_discrete_log_pollard_rho` do not take `prime_order` as an argument.

---

I would like to clearly state in the docstring the conditions for the arguments to be given to `_discrete_log_index_calculus`. The `order` must be a prime number, but I don't think that is sufficient. The `order` must be somewhat large to be calculated. Also, if `relation[index]` is a multiple of `order`, we cannot calculate `rinv = pow(relation[index],-1,order)`.

---

You can use `else` to write without using `index`.
https://docs.python.org/3/reference/compound_stmts.html#the-for-statement

```python
        for i in range(lf):
            ...
            if relationa[i] > 0 and index == lf:
                index = i
                break
        if index == lf:
            return relationa[lf] * (order - 1) % order
```

```python
        for i in range(lf):
            ...
            if relationa[i] > 0:
                break
        else:
            return relationa[lf] * (order - 1) % order
```

**Comment by teschlg** [comment]
> I see. Then let's just go with it. But is there a high probability that `n` will be smooth? If it is high, it would be better to use `n ! = 1` rather than checking `if n == 1: break` every time. Maybe the probability is low, so we omit `n ! = 1` can be omitted from the check.
This seems indeed a bit faster. Done.


> > I have added a reference for how to obtain B.
> 
> Thanks, I wasn't sure where the 0.7 was coming from, but 0.5 is correct (Newton's method is probably wrong too). While several approximate values are used to find the optimal `B`, there is not much advantage to be exact only for `u`. Thus, I see no problem in finding `B` directly.
> 
The 0.7 is a heuristic I have seen others using. The asymptotic value is not good in the regime we can handle. My current approach gave the best performance in my test.

> > The `prime_order` test is just how it is done in the other functions. I just tried to be consistent here.
> 
> Consistent with what? `_discrete_log_shanks_steps` and `_discrete_log_pollard_rho` do not take `prime_order` as an argument.
> 
I meant, that the factorization is passed along to avoid another factorizations. So I felt it is appropriate to pass the information that the order is prime to avoid a useless primality test.


> I would like to clearly state in the docstring the conditions for the arguments to be given to `_discrete_log_index_calculus`. The `order` must be a prime number, but I don't think that is sufficient. The `order` must be somewhat large to be calculated.
I do not understand this last line.
> Also, if `relation[index]` is a multiple of `order`, we cannot calculate `rinv = pow(relation[index],-1,order)`.
Why, the entries are all taken modulo the order and if the order is prime we have a field. So every nonzero entry has an inverse.


> You can use `else` to write without using `index`. https://docs.python.org/3/reference/compound_stmts.html#the-for-statement
Done.

**Comment by haru-44** [comment]
> I meant, that the factorization is passed along to avoid another factorizations. So I felt it is appropriate to pass the information that the order is prime to avoid a useless primality test.

For example, `_primitive_root_prime_iter` expects `p` to be prime, but does not check whether it is prime. Private functions omit input checks because they are not expected to be used by public users. On the other hand, it is necessary to include appropriate descriptions in the comments to make it clear to sympy developers what parameters may (or may not) be given.

> The asymptotic value is not good in the regime we can handle. My current approach gave the best performance in my tests.

Even if the resulting performance is better, the approach is not the same. I would prefer to adjust the constant `c` for good performance as `B = c * exp(0.5 * sqrt(log(n)*log(n))))`.

We would like to choose `B` to minimize the computational cost, but we are discarding a lot when estimating the computational cost. Therefore, it is not uncommon for the theoretically obtained `B` to differ from the actual optimal `B`. Hence, "see the pdf for the theory that `B` is $L_n[1/2, 1/2]$. However, we adjust `c` for performance" is a satisfactory explanation.

**Comment by teschlg** [comment]
I see no theoretical justification for adding the suggested constant c. Consequently, there is no choice for c which will give good performance for all p. Moreover, the correct asymptotic value is not good approximation for the range of p we would like to handle. I have changed the value by a heuristic correction to the asymptotic value which works reasonable for p < 2^70

**Comment by haru-44** [comment]
I apologize for the delayed response.

Please include a seed as an argument to eliminate randomness. Refer to the `_discrete_log_pollard_rho` function for an example.

Please clarify what values may be given to `_discrete_log_index_calculus`. Even if `order` is a prime number, it may cause an error. For example, `_discrete_log_index_calculus(5, 4, 4, 2)` raises a `ValueError`. `_discrete_log_index_calculus(18, 7, 7, 3)` will not stop.

**Comment by teschlg** [comment]
Thanks for your feedback!

I hope I got the "seed" right.

The second issue seems more tricky. The problem is, that certain primes in the factorbase don't have a log. I could not find anything concerning this in the literature I checked. I tried to remove those primes from the factorbase or replace them by a power such that the log exists. Both options gave a significant slowdown. It seems the errors always cancel (but I did not try to prove this rigorously). To be sure I have added a check that the solution works. I have also limited the number of iterations in case things take significantly longer than expected.

**Comment by haru-44** [comment]
> I hope I got the "seed" right.

Thanks

> The second issue seems more tricky. The problem is, that certain primes in the factorbase don't have a log. I could not find anything concerning this in the literature I checked. I tried to remove those primes from the factorbase or replace them by a power such that the log exists. Both options gave a significant slowdown. It seems the errors always cancel (but I did not try to prove this rigorously). To be sure I have added a check that the solution works. I have also limited the number of iterations in case things take significantly longer than expected.

I thought it would not occur if `order` was sufficiently large. I simply wanted it to be explicitly stated if there is a specific condition (isn't `3*sqrt(log(n)*log(log(n))) < log(order)` that condition?).

- If `relation = [r % order for r in relation]`, then `rinv = pow(relation[index],-1,order)` should not raise an error. We can do the same thing with `_discrete_log_is_smooth`.
- Obviously, if `order` is less than `lf`, there are not enough variations of `x = randint(0,ordermo)`.

**Comment by teschlg** [comment]
> I thought it would not occur if `order` was sufficiently large. 

Yes, it will not occur in situations where the algorithm makes sense to use. I still added `%m` as to not raise any exceptions.

>I simply wanted it to be explicitly stated if there is a specific condition (isn't `3*sqrt(log(n)*log(log(n))) < log(order)` that condition?).

This is the heuristic condition, where Index Calculus should be faster than BSGS or RHO.
> * If `relation = [r % order for r in relation]`, then `rinv = pow(relation[index],-1,order)` should not raise an error. We can do the same thing with `_discrete_log_is_smooth`.

Yes, but somehow I felt it is more appropriate to do it in the main part of the function.
> * Obviously, if `order` is less than `lf`, there are not enough variations of `x = randint(0,ordermo)`.

Yes, but this will be caught eventually and there is a chance that it still might work. So why give up too early?
In any case, index calculus will not be invoked by `discrete_log` if `order` is small.

**Comment by haru-44** [comment]
My thoughts on this issue are to leave the following comments on the [previous version](https://github.com/sympy/sympy/pull/26488/commits/df6f7c5f57832ef0410e4e5007420e5c4aa75f69).

- That `order` must be a prime number and must be somewhat large ( not sure of the exact conditions for that).
- That we are using `3*sqrt(log(n)*log(log(n))) < log(order)`, which is a heuristic condition.

The important thing is not to confuse the sympy developer who will be looking at `_discrete_log_index_calculus` next.

File: sympy/ntheory/residue_ntheory.py [hunk 1]
```diff
@@ -1432,6 +1433,121 @@ def _discrete_log_pollard_rho(n, a, b, order=None, retries=10, rseed=None):
...
+        if tmp == 1:
+            return x * (order - 1) % order
+        relationa = _discrete_log_is_smooth(tmp, factorbase)
+        if relationa:
+            for i in range(lf):
+                relationa[i] %= order
+            relationa += [ x ]
+            break
+        bx = bx * b % n
+        x += 1
```
Comment (haru-44) [inline sympy/ntheory/residue_ntheory.py:1502]:
```suggestion
    abx = 1
    for x in range(order):
        abx = abx * b % n # abx = a*pow(b, x, n) % n
        if abx == 1:
            return x * (order - 1) % order
        relationa = _discrete_log_is_smooth(abx, factorbase)
        if relationa:
            relationa = [r % order for r in relationa] + [x]
            break
```

File: sympy/ntheory/residue_ntheory.py [hunk 2]
```diff
@@ -1432,6 +1433,121 @@ def _discrete_log_pollard_rho(n, a, b, order=None, retries=10, rseed=None):
...
+    else:
+        raise ValueError("Index Calculus failed")
+
+    relations = [None] * lf
+    k=1  # number of relations found
```
Comment (haru-44) [inline sympy/ntheory/residue_ntheory.py:1507]:
```suggestion
    k = 1  # number of relations found
```

File: sympy/ntheory/residue_ntheory.py [hunk 3]
```diff
@@ -1432,6 +1433,121 @@ def _discrete_log_pollard_rho(n, a, b, order=None, retries=10, rseed=None):
...
+        index = lf  # determine the index of the first nonzero entry
+        for i in range(lf):
+            ri = relation[i] % order
+            if ri> 0 and relations[i] is not None:  # make this entry zero if we can
+                for j in range(lf+1):
+                    relation[j] = (relation[j] - ri*relations[i][j]) % order
+            else:
+                relation[i] = ri
+            if relation[i] > 0 and index == lf:  # is this the index of the first nonzero entry?
+                index= i
```
Comment (haru-44) [inline sympy/ntheory/residue_ntheory.py:1527]:
```suggestion
                index = i
```

File: sympy/ntheory/residue_ntheory.py [hunk 4]
```diff
@@ -1432,6 +1433,121 @@ def _discrete_log_pollard_rho(n, a, b, order=None, retries=10, rseed=None):
...
+                    relationa[j] = (relationa[j] - rbi*relations[i][j]) % order
+            if relationa[i] > 0:  # the index of the first nonzero entry
+                break  # we do not need to reduce further at this point
+        else:  # all unkowns are gone
+            #print(f"Success after {k} relations out of {lf}")
+            x = relationa[lf] * (order - 1) % order
+            if pow(b,x,n) == a:
+                return x
+            else:
+                raise ValueError("Index Calculus failed")
```
Comment (haru-44) [inline sympy/ntheory/residue_ntheory.py:1548]:
```suggestion
            raise ValueError("Index Calculus failed")
```

File: sympy/ntheory/tests/test_residue.py [hunk 5]
```diff
@@ -265,6 +269,8 @@ def test_residue():
     assert discrete_log(2456747, 3**51, 3) == 51
     assert discrete_log(32942478, 11**127, 11) == 127
     assert discrete_log(432751500361, 7**324, 7) == 324
+    assert discrete_log(265390227570863,184500076053622, 2) == 17835221372061
+    assert discrete_log(22708823198678103974314518195029102158525052496759285596453269189798311427475159776411276642277139650833937, 17463946429475485293747680247507700244427944625055089103624311227422110546803452417458985046168310373075327, 123456) == 2068031853682195777930683306640554533145512201725884603914601918777510185469769997054750835368413389728895
```
Comment (haru-44) [inline sympy/ntheory/tests/test_residue.py:273]:
```suggestion
    assert discrete_log(22708823198678103974314518195029102158525052496759285596453269189798311427475159776411276642277139650833937,
                        17463946429475485293747680247507700244427944625055089103624311227422110546803452417458985046168310373075327,
                        123456) == 2068031853682195777930683306640554533145512201725884603914601918777510185469769997054750835368413389728895
```

File: sympy/ntheory/residue_ntheory.py [hunk 6]
```diff
@@ -1432,6 +1433,121 @@ def _discrete_log_pollard_rho(n, a, b, order=None, retries=10, rseed=None):
...
+    raise ValueError("Index Calculus failed")
+
```
Comment (haru-44) [inline sympy/ntheory/residue_ntheory.py:1550]:
```suggestion


```

File: sympy/ntheory/residue_ntheory.py [hunk 7]
```diff
@@ -1432,6 +1433,121 @@ def _discrete_log_pollard_rho(n, a, b, order=None, retries=10, rseed=None):
...
 
 
+def _discrete_log_is_smooth(n: int, factorbase: list):
+    """Try to factor n with respect to a given factorbase.
+    Upon success a list of exponents with repect to the factorbase is returned.
+    Otherwise None."""
+    factors = [0]*len(factorbase)
+    for i, p in enumerate(factorbase):
+        while n % p == 0: # divide by p as many times as possible
+            factors[i] +=1
```
Comment (haru-44) [inline sympy/ntheory/residue_ntheory.py:1443]:
```suggestion
            factors[i] += 1
```

File: sympy/ntheory/residue_ntheory.py [hunk 8]
```diff
@@ -1550,11 +1668,15 @@ def discrete_log(n, a, b, order=None, prime_order=None):
...
-            return _discrete_log_shanks_steps(n, a, b, order)
-        return _discrete_log_pollard_rho(n, a, b, order, order_factors)
-
-    return _discrete_log_pohlig_hellman(n, a, b, order)
+        # Shanks and Pollard rho are O(sqrt(order)) while index calculus is O(exp(2*sqrt(log(n)log(log(n)))))
+        # we compare the expected running times to determine the algorithmus which is expected to be faster
+        if 3*sqrt(log(n)*log(log(n))) < log(order):  # the number 3 was determined experimental
+            return _discrete_log_index_calculus(n, a, b, order)
+        elif order < 1000000000000:
+            return _discrete_log_shanks_steps(n, a, b, order)  # Shanks seems typically faster, but uses O(sqrt(order)) memory
```
Comment (haru-44) [inline sympy/ntheory/residue_ntheory.py:1676]:
```suggestion
            # Shanks seems typically faster, but uses O(sqrt(order)) memory
            return _discrete_log_shanks_steps(n, a, b, order)
```

In [PEP 8](https://peps.python.org/pep-0008/#maximum-line-length), it is recommended that each line should contain no more than 79 characters. While SymPy does not strictly enforce this rule, it is advisable to adhere to it when possible.

File: sympy/ntheory/residue_ntheory.py (hunk 1 above)
Comment (teschlg) [reply sympy/ntheory/residue_ntheory.py:1502]: Did you test this? In principle I agree, but it seems to me we need to start with abx = a.

**Comment by teschlg** [comment]
I have added all your suggestions (and fixed one of them;-). Thanks!

File: sympy/ntheory/residue_ntheory.py (hunk 1 above)
Comment (haru-44) [reply sympy/ntheory/residue_ntheory.py]: You're right.

**Comment by haru-44** [comment]
I think `(order - x) % order` should be `order - x`

**Comment by teschlg** [comment]
> I think `(order - x) % order` should be `order - x`

Unless `x == 0`

**Review by haru-44 — APPROVED** [review]
Thank you for your patience.
If there are no other suggestions, I will merge them.

**Comment by teschlg** [comment]
Thanks for your suggestions! I went over everything once more and found one stupid mistake. Since I take logs when comparing the run times, the constant must be additive and not multiplicative.

**Pull request merged**